src/test/java/htsjdk/samtools/BAMSBIIndexerTest.java

```diff
@@ -0,0 +1,157 @@
+                .validationStringency(ValidationStringency.SILENT)
+                .enable(SamReaderFactory.Option.INCLUDE_SOURCE_IN_RECORDS).open(LARGE_BAM_FILE);
+        final List<SAMRecord> allReads = Iterables.slurp(samReader);
+
+        final List<SAMRecord> allReadsFromChunks = new ArrayList<>();
+        for (final Chunk chunk : chunks) {
+            allReadsFromChunks.addAll(getReadsInChunk(LARGE_BAM_FILE, chunk));
+        }
+        Assert.assertEquals(allReadsFromChunks, allReads);
+
+        final List<Chunk> optimizedChunks = Chunk.optimizeChunkList(chunks, 0);
+        Assert.assertEquals(optimizedChunks.size(), 1);
+        final List<SAMRecord> allReadsFromOneChunk = getReadsInChunk(LARGE_BAM_FILE, optimizedChunks.get(0));
+        Assert.assertEquals(allReadsFromOneChunk, allReads);
+    }
+
+    @Test
+    public void testIndexersProduceSameIndexes() throws Exception {
+        final long bamFileSize = BAM_FILE.length();
+        for (final long g : new long[] { 1, 2, 10, SBIIndexWriter.DEFAULT_GRANULARITY }) {
+            final SBIIndex index1 = fromBAMFile(BAM_FILE, g);
+            final SBIIndex index2 = fromSAMRecords(BAM_FILE, g);
+            Assert.assertEquals(index1, index2);
+            Assert.assertEquals(index1.dataFileLength(), bamFileSize);
+            Assert.assertEquals(index2.dataFileLength(), bamFileSize);
+        }
+    }
+
+    private SBIIndex fromBAMFile(final File bamFile, final long granularity) throws IOException {
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        BAMSBIIndexer.createIndex(new SeekableFileStream(bamFile), out, granularity);
+        return SBIIndex.load(new ByteArrayInputStream(out.toByteArray()));
+    }
+
+    private SBIIndex fromSAMRecords(final File bamFile, final long granularity) {
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        final SBIIndexWriter indexWriter = new SBIIndexWriter(out, granularity);
+        final BAMFileReader bamFileReader = bamFileReader(bamFile);
+        final CloseableIterator<SAMRecord> iterator = bamFileReader.getIterator();
+        while (iterator.hasNext()) {
+            processAlignment(indexWriter, iterator.next());
+        }
+        indexWriter.finish(bamFileReader.getVirtualFilePointer(), bamFile.length());
+        return SBIIndex.load(new ByteArrayInputStream(out.toByteArray()));
+    }
+
+    public void processAlignment(final SBIIndexWriter indexWriter, final SAMRecord rec) {
+        final SAMFileSource source = rec.getFileSource();
+        if (source == null) {
+            throw new SAMException("No source (virtual file offsets); needed for indexing on BAM Record " + rec);
+        }
+        final BAMFileSpan filePointer = (BAMFileSpan) source.getFilePointer();
+        indexWriter.processRecord(filePointer.getFirstOffset());
+    }
+
+    private BAMFileReader bamFileReader(final File bamFile) {
+        final SamReader samReader = SamReaderFactory.makeDefault()
```

Comment:
The reader is used outside the scope of this method (but within the test since it's private), so it cannot be closed in a TWR block. All usages of this method use a TWR though, which is what is important.

Reply:
TIL thanks.